### PR TITLE
Support for ALTER VITESS_MIGRATION statements

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

This PR introduces the following queries:

- `ALTER VITESS_MIGRATION '8c096d5e_81c0_11eb_a8d2_f875a4d24e90' RETRY`
- `ALTER VITESS_MIGRATION '8c096d5e_81c0_11eb_a8d2_f875a4d24e90' CANCEL`
- `ALTER VITESS_MIGRATION CANCEL ALL`

The above map to these `vtctl` commands, respectively:

- `vtctl OnlineDDL <keyspace> retry 8c096d5e_81c0_11eb_a8d2_f875a4d24e90`
- `vtctl OnlineDDL <keyspace> cancel 8c096d5e_81c0_11eb_a8d2_f875a4d24e90`
- `vtctl OnlineDDL <keyspace> cancel all`

Also, reserved but not implemented yet:
- `ALTER VITESS_MIGRATION '8c096d5e_81c0_11eb_a8d2_f875a4d24e90' COMPLETE`

Returned result set for these queries is always empty, but `AffectedRows` indicates how many migrations were actually affected by the command.

The various `endtoend/onlineddl` tests now use these queries.

## Related Issue(s)

- #7604 
- #6926 

## Checklist
- [ ] Should this PR be backported?
- [x] Tests were added or are not required
- [ ] Documentation was added or is not required

## Impacted Areas in Vitess
Components that this PR will affect:

- [x]  Query Serving
- [ ]  VReplication
- [ ]  Cluster Management
- [ ]  Build/CI
- [ ]  VTAdmin

### CODE REVIEW

**Comment by shlomi-noach** [comment]
This PR extends https://github.com/vitessio/vitess/pull/7656 ; it makes sense to review this PR after https://github.com/vitessio/vitess/pull/7656 is merged.

**Comment by shlomi-noach** [comment]
Documentation to follow

**Comment by shlomi-noach** [comment]
This PR also refactors `endtoend/onlineddl` tests by consolidating much of the shared logic and refactoring functions into `endtoend/onlineddl/vtgate_utils.go`

**Comment by shlomi-noach** [comment]
@systay points out that this PR has commits from merged PRs. It seems to be a rebase gone wrong. However, code-wise and diff-wise there doesn't seem to be a risk here.

**Comment by shlomi-noach** [comment]
Confused about `endtoend / End-to-End Test` failures, e.g.: https://github.com/vitessio/vitess/pull/7663/checks?check_run_id=2105797750

They're consistently failing after `5m`. It seems like a timeout issue. But I can't find what times out after `5m`. The test itself (https://github.com/vitessio/vitess/blob/master/.github/workflows/endtoend.yml) is set with `timeout-minutes: 30`

**Comment by shlomi-noach** [comment]
Kind request for review :pray: (reminder: this PR extends #7656)

**Comment by shlomi-noach** [comment]
will resolve conflicts after #7656 is merged. The conflicts are due to #7669. See https://github.com/vitessio/vitess/pull/7656/commits/d0c741f5175d25384bc02611165e35070f554453 for how conflict was resolved in #7656

**Comment by shlomi-noach** [comment]
Sheesh. GitHub Actions is having another bad day.

File: go/vt/vtgate/planbuilder/migration.go [hunk 1]
```diff
@@ -17,15 +17,39 @@ limitations under the License.
...
 	topodatapb "vitess.io/vitess/go/vt/proto/topodata"
-
 	vtrpcpb "vitess.io/vitess/go/vt/proto/vtrpc"
 	"vitess.io/vitess/go/vt/sqlparser"
 	"vitess.io/vitess/go/vt/vterrors"
 	"vitess.io/vitess/go/vt/vtgate/engine"
 )
 
-func buildRevertPlan(query string, stmt *sqlparser.RevertMigration, vschema ContextVSchema) (engine.Primitive, error) {
+func buildAlterMigrationPlan(query string, stmt *sqlparser.AlterMigration, vschema ContextVSchema) (engine.Primitive, error) {
```
Comment (systay) [inline go/vt/vtgate/planbuilder/migration.go:28]: nit: why ask for `stmt` and then not use it?

File: go/vt/vttablet/onlineddl/executor.go [hunk 2]
```diff
@@ -2236,10 +2236,11 @@ func (e *Executor) updateMigrationProgress(ctx context.Context, uuid string, pro
 	return err
 }
 
-func (e *Executor) retryMigration(ctx context.Context, whereExpr string) (result *sqltypes.Result, err error) {
+// retryMigrationWhere retries a migration based on a given WHERE clause
+func (e *Executor) retryMigrationWhere(ctx context.Context, whereExpr string) (result *sqltypes.Result, err error) {
 	e.migrationMutex.Lock()
 	defer e.migrationMutex.Unlock()
-	parsed := sqlparser.BuildParsedQuery(sqlRetryMigration, ":tablet", whereExpr)
+	parsed := sqlparser.BuildParsedQuery(sqlRetryMigrationWhere, ":tablet", whereExpr)
```
Comment (systay) [inline go/vt/vttablet/onlineddl/executor.go:2243]: just noticed that three lines down, on line 2247, we are never checking err from `parsed.GenerateQuery`

File: go/vt/vtgate/planbuilder/migration.go (hunk 1 above)
Comment (shlomi-noach) [reply go/vt/vtgate/planbuilder/migration.go:28]: Nice catch! Removed.

File: go/vt/vttablet/onlineddl/executor.go (hunk 2 above)
Comment (shlomi-noach) [reply go/vt/vttablet/onlineddl/executor.go:2243]: Fixed!

**Review by systay — APPROVED** [review]
LGTM!

**Pull request merged**